### PR TITLE
feat(140): add `--signalEmittedFiles` and trigger `file_emitted` silently

### PR DESCRIPTION
**Pull request by pp0rtal**
In #140 I've added support to new event `.on(`file_emitted`)`. This feature works as long `--listEmittedFiles` param is set.
Problem: `--listEmittedFiles`is noisy and print a bunch of `TSFILE: /path/to/emitted/files.js`
@gilamran have suggested a new param `--signalEmittedFiles` which would make the event work, but silently.

### What's here
- parse `--signalEmittedFiles` (boolean)
- make it call tsc with `--listEmittedFiles`
- make it hide ines starting with `TSFILE:`

### CODE REVIEW

File: src/lib/stdout-manipulator.ts [hunk 1]
```diff
@@ -54,7 +54,10 @@ function color(line: string, noClear: boolean = false): string {
   return line;
 }
 
-export function print(noColors: boolean, noClear: boolean, line: string): void {
+export function print(noColors: boolean, noClear: boolean, line: string, signalEmittedFiles: boolean): void {
```
Comment (gilamran) [inline src/lib/stdout-manipulator.ts:57]: maybe it's time to move `line: string` to be the first parameter?

File: src/lib/stdout-manipulator.ts [hunk 2]
```diff
@@ -54,7 +54,10 @@ function color(line: string, noClear: boolean = false): string {
...
+  const outputLine = noColors ? line : color(line, noClear);
```
Comment (gilamran) [inline src/lib/stdout-manipulator.ts:58]: I think that the content of the function should be like this:
```
if (signalEmittedFiles && outputLine.startsWith('TSFILE:')) { 
  return; 
}
console.log(noColors ? line : color(line, noClear));
```

**Review by gilamran — COMMENTED** [review]
I think that you shouldn't remove the TSFILE lines if the user explicitly added `--listEmittedFiles` to the args.

File: src/lib/stdout-manipulator.ts (hunk 1 above)
Comment (pp0rtal) [reply src/lib/stdout-manipulator.ts]: @gilamran Definitively! What do you think of `print(lint, { optionalParam: bool })` format? 81c3159fed10c89b3bbb1a2d653f5f129237fa00
I can remove the ability to make them optional if you prefer as they are explicit in the only usage.

File: src/lib/stdout-manipulator.ts (hunk 2 above)
Comment (pp0rtal) [reply src/lib/stdout-manipulator.ts]: @gilamran Definitively, I started to introduce `outputLine` but actually we don't need it (we need the raw line) it was fixed in the same commit 81c3159fed10c89b3bbb1a2d653f5f129237fa00 good catch

**Comment by pp0rtal** [comment]
> I think that you shouldn't remove the TSFILE lines if the user explicitly added `--listEmittedFiles` to the args.

I understand, those params are complementary ;
- either the lib user calls `tsc-watch.js --listEmittedFiles`, and in this case he doesn't need the `--signalEmittedFiles`
(wants `emitted_file` signals but not the output)
- either he use `tsc-watch.js --signalEmittedFiles` to make events work, but then shouldn't add `--signalEmittedFiles`
(wants `emitted_file` signals and the output)

Do you think it's worth checking if `listEmittedFiles` is inside tsc `args`?

**Comment by gilamran** [comment]
> Do you think it's worth checking if `listEmittedFiles` is inside tsc `args`?
Yes, I think that it's more clear

File: src/lib/stdout-manipulator.ts [hunk 3]
```diff
@@ -54,7 +54,23 @@ function color(line: string, noClear: boolean = false): string {
   return line;
 }
 
-export function print(noColors: boolean, noClear: boolean, line: string): void {
+type PrintParams = {
```
Comment (gilamran) [inline src/lib/stdout-manipulator.ts:57]: please rename to `TPrintParams`

Comment (pp0rtal) [reply src/lib/stdout-manipulator.ts]: @gilamran Done in 9f88210c206a9b2af5e4ec0e47dfa3543bd45555 (commit to squash)

**Comment by pp0rtal** [comment]
@gilamran Then I'll switch signalEmittedFiles virtually off c8a34c0bdd7716de7690f5123d37fd6d85513330 if `--listEmittedFiles` is found
I initially implemented in `tsc-watch.ts` file but it was too complex (and test uncovered) for nothing, hope you'll find the solution convenient.

File: src/lib/args-manager.ts [hunk 4]
```diff
@@ -54,6 +54,7 @@ export function extractArgs(inputArgs: string[]) {
   const noColors = extractCommand(args, '--noColors');
   const noClear = extractCommand(args, '--noClear');
   const silent = extractCommand(args, '--silent');
+  const signalEmittedFiles = extractCommand(args, '--signalEmittedFiles') && !args.includes('--listEmittedFiles');
```
Comment (gilamran) [inline src/lib/args-manager.ts:57]: I think that we have some naming problem here...
maybe this `const` should be called `dontPrintTSFILE`?

Comment (pp0rtal) [reply src/lib/args-manager.ts:57]: @gilamran More names, more complexity, if `--signalEmittedFiles` is specified, I'd rather keep `signalEmittedFiles`
`dontPrintTSFILE` looks surprising for a parser, no?
The role of `&& !args.includes('--listEmittedFiles');` is to virtually make as if `--signalEmittedFiles` was not specified.

Comment (gilamran) [reply src/lib/args-manager.ts:57]: It's a bit confusing that we do send signals for emitted file if `--listEmittedFiles` is specificed. and `--signalEmittedFiles` meaning is actually "Only signal emitted file, don't print them".
If you can think of a better name, that would be great. if not that's also ok...

Comment (pp0rtal) [reply src/lib/args-manager.ts:57]: @gilamran I tried with your wording, but I'm really annoyed when if comes to this in `tsc-watch.ts` for the dual usage (both call tsc with right args, and call print) :confused: 
```typescript
  if (dontPrintTSFILE) {
    nodeArgs.push(`--listEmittedFiles`);
  }
```

The variable could be named `listEmittedFilesOrHideTSFILE` to match all needs
but hiding this complexity in the parser will potentially make this more maintainable.

I think I will stop here, however thank you for the suggestion

Comment (gilamran) [reply src/lib/args-manager.ts:57]: I understand, what about we'll have 2 variables one that extract the `signalEmittedFiles` and one that indicates if `--listEmittedFiles` was requested. Lets call the 2nd one `requestedToListEmittedFiles`.
In this case we can do the `if` like this:
```
if (signalEmittedFiles && !requestedToListEmittedFiles) {
  return;
}
```

File: src/lib/tsc-watch.ts [hunk 5]
```diff
@@ -84,13 +85,17 @@ function getTscPath(): string {
...
+  signalEmittedFiles: boolean;
 }
 
-function spawnTsc({ maxNodeMem }: INodeSettings, args: string[]): ChildProcess {
+function spawnTsc({ maxNodeMem, signalEmittedFiles }: INodeSettings, args: string[]): ChildProcess {
   const nodeArgs = [];
   if (maxNodeMem) {
     nodeArgs.push(`--max_old_space_size=${maxNodeMem}`);
   }
+  if (signalEmittedFiles) {
```
Comment (gilamran) [inline src/lib/tsc-watch.ts:96]: This should also changed to this:
```
if (signalEmittedFiles && !requestedToListEmittedFiles) {
  nodeArgs.push(`--listEmittedFiles`);
}
}
```

**Comment by gilamran** [comment]
Both merged. thanks for all the work you've done!

**Pull request merged**